### PR TITLE
Integrate DL1DataHandler

### LINKED ISSUE #40: Add option to list data files directly in configuration file
Since YAML provides a convenient way to include lists directly in the config file, allow Data:file_list to accept either a path to a file containing file paths (the current method) or a list of file paths written directly in the config file.

### LINKED ISSUE #41: Clean up input_fn in run_model.py
Use dictionary unpacking to pass in settings instead of passing the dictionary directly. This will provide a convenient way to provide default arguments, making it easier to have optional data input settings. At least one options that is currently configurable should be hardcoded, that is whether to use dataset.map() - this is already constrained by the choice of data format. Also, investigate whether it's worth making whether to prefetch a configurable option or if there's a best practice for it that could just be hardcoded.

### LINKED ISSUE #46: Define API for accessing DataLoader class attributes
All HDF5DataLoader class attributes are currently specific to that class as opposed to DataLoader, but are accessed outside the class in run_model.py. 

When defining data_loader, `generator_output_dtypes`, `map_fn_output_dtypes`, `output_names`, and `output_is_label` are all accessed. It doesn't seem these are specific to the data format, so if they are made part of the DataLoader base class, then this entire section except for `data_loader = HDF5DataLoader()` can be pulled out of the if clause, resulting in cleaner code. 

Similarly, when getting the event indices in predict mode, `example_type` and `examples` are accessed and the event index names are manually defined, but again it may be possible to define all these in such a way that the event indices whatever they may be could be accessed for any DataLoader without relying on specific attributes for each DataLoader subclass.

This issue is probably best tackled at the time we actually add another DataLoader subclass.

### LINKED ISSUE #73: Function that takes a loading configuration and returns a dataloader
Possible improvement suggested by @aribrill in pull request #71 

> I wonder then if we should create a common function that can be loaded by different scripts that takes in a config Data section and returns a DataLoader.

### LINKED ISSUE #82: Define API for DataLoader output
With the addition of multiple telescope/camera types and need to include more types of auxiliary information, the API for DataLoader should be more explicitly defined. There are three kinds of image data to return: 1) single telescope, 2) array of one telescope type with triggers, and 3) array of multiple telescope types with triggers. The practical difference between case 2 and case 3 just using one tel type is that the name of the data array may be generic in case 2 (`telescope_data`) but not in case 3. The removal of case 2 in #74 is the cause of the difficulty noted in #81, and it should be restored.

In addition, there are two kinds of auxiliary data: telescope-based (anything included in the [ImageExtractor Array Info table](https://github.com/cta-observatory/image-extractor/wiki/CTA-ML-Data-Format#array_info-table), in practice just positions) and event-based (everything included in the [ImageExtractor Event Info table](https://github.com/cta-observatory/image-extractor/wiki/CTA-ML-Data-Format#event_info-table)). Note that any information used as event-based auxiliary data could be used as a label and vice-versa.

The telescope positions are currently normalized in DataLoader, but this behavior is processing which may not always be desired. This functionality should be moved to DataProcessor and made a configuration option.

Then, the following configuration parameters in Data:Loading need to added, _changed_, or ~~removed~~ (the parameter names are open to change):
- _example_type_: single_tel, array_one_tel_type, array_multiple_tel_types
- _selected_tel_types_: tel_type for single_tel or array_one_tel_type, or list of tel_types for array_multiple_tel_types
- telescope_auxiliary_data: None, positions
- event_auxiliary_data: None or a list of desired auxiliary data
- labels: None or a list of desired labels
- ~~merge_tel_types~~: confusing and unnecessary to support

All combinations are valid and should be supported except for single tel examples with telescope auxiliary data. At least the following event auxiliary data / labels should be supported: particle id (aka gamma_hadron_label), MC energy (for signal efficiency #58 and energy reconstruction #67), event zenith and azimuth (angular reconstruction #67).

DataLoader should return a structure describing the examples, mapping the index of each output array to its name, shape (from DataProcessor), dtype, and output type (data or label). This structure should be included in the metadata, where it would subsume some current parameters such as `image_shapes` and `total_aux_params`. It would be used both by models to parse the data arrays and by `run_model.py` to construct the input_fn, eliminating the need to access DataLoader attributes directly (#46). It may also make sense to eliminate the DataLoader.get_auxiliary_data() method, since it would be somewhat ambiguous, and replace it with a method to explicitly return the telescope positions.

### LINKED ISSUE #115: Use DL1DataHandler for data loading
Replace the classes currently used for data loading (DataLoader, DataProcessor, and ImageMapper) with the updated version in the [DL1DataHandler](https://github.com/cta-observatory/dl1-data-handler) package. This package was designed to fix the problems noted in and so resolve #82. It also fixes #46 and closes #73 as no longer needed. This issue involves the follow steps:

- [x] Remove data_loading.py, data_processor.py, image_mapping.py, and pixel_pos_files, which have been superseded by DL1DH.
- [x] Remove all scripts and notebooks that exclusively use or are designed to test these modules and create issues in the DL1DH repository to recreate them there as needed.
- [x] Update run_model.py to use DL1DataReader instead of DataLoader and DataProcessor to load and process the data, specifically using the new DL1DataReader.example_description to more cleanly define the input_fn (#46).
- [x] Update run_model.py to use the capability of DL1DataReader to load any telescope and event info to:
  - [x] Load arbitrary auxiliary info,
  - ~~Define and calculate the loss for multiple labels~~, and
  - [x] Simplify making predictions by returning the true label and event ids within the predictions dictionary
- [x] Update the configuration arguments where needed to match the DL1DH API.
- [x] Update to reflect these changes:
  - [x] README.md
  - [x] example_config.yml
  - [x] MANIFEST.in
- [ ] Do a benchmark run to confirm that the changes do not affect speed and performance.

### PR DESCRIPTION
**Pull request by aribrill**
This PR integrates DL1DataHandler into CTLearn, superseding the old DataLoader, DataProcessor, and ImageMapper. Although the core of CTLearn remains the same, this integration necessitates significant changes to the API, as the configuration arguments have been changes to match those taken by DL1DataReader. In addition, since DL1DataReader adds new features not available with DataLoader, data input, predicting, and logging information about the dataset have been simplified and improved as well.

This PR closes #115, which will in turn resolve #82, fix #46, and close #73. It will incidentally also fix #40 and fix #41 .

The following still need to be resolved before this PR can be merged (see also #115):

- [x] Merge DL1DataHandler PRs cta-observatory/dl1-data-handler/pull/46 and cta-observatory/dl1-data-handler/pull/50
- [x] Solve incompatibility in Python version requirement between CTLearn (3.6.5) and DL1DataHandler (>=3.7.0)
 - [x] Do a benchmark run to verify that the changes do not decrease speed and performance

### CODE REVIEW

**Comment by Jsevillamol** [comment]
The names of the telescopes in the multiple_configurations section of the example_configuration.yaml are incorrect, right?

**Comment by aribrill** [comment]
> The names of the telescopes in the multiple_configurations section of the example_configuration.yaml are incorrect, right?

Correct, fixed.
